Changes for TimescaleDB 2.0.0-rc2 launch (#543)
timescale/docs.timescale.com-content: getting-started/release-notes/changes-in-timescaledb-2.md

# Changes in TimescaleDB 2.0

TimescaleDB 2.0 introduces new features and capabilities to its advanced relational database for time-series data. Driven by user feedback and our experience building products like [Promscale](https://github.com/timescale/promscale), 2.0 is a major milestone and introduces the first multi-node, petabyte-scale relational database for time-series. In addition to multi-node capabilities, this release includes new features - and improvements to existing ones - focused on giving users more flexibility, control over their data and the ability to customize behavior to suit their needs.

To facilitate many of the improvements in [TimescaleDB 2.0](https://github.com/timescale/timescaledb/releases/tag/2.0.0-rc1), several existing APIs and function definitions have been modified which may require updates to your existing code. 

Most notably, the following features include updates and changes that may impact existing code using these interfaces. If your workloads use any of the features, this document covers each in detail to help you understand what - if any - action you need to take.

> *   **Dropping support for PostgreSQL versions 9.6 and 10:** As mentioned in [our upgrade documentation](https://docs.timescale.com/update-timescaledb), TimescaleDB 2.0 no longer supports older PostgreSQL versions. You will need to be running PostgreSQL version 11 or 12 to upgrade your TimescaleDB installation to 2.0.

Author: Reworded slightly

*   **Informational Views:** Based on user feedback and our desire to make TimescaleDB easy to manage, multiple information views have been consolidated for better clarity. 

The following migration document provides more information on each of these API changes and how they affect users of TimescaleDB 1.x . We’ve also included a bit of information about the impetus for these changes and our design decisions. For a more in-depth coverage of APIs, as well as a detailed description of new capabilities included in 2.0, such as distributed hypertables and the ability to run user-defined actions (custom background jobs), please [review our updated documentation](https://docs.timescale.com/v2.0/main) (navigate to specific features of interest).

Once you have read through this guide and understand the impact upgrading to the latest version may have on your existing application and infrastructure, please follow our [upgrade to TimescaleDB 2.0](https://docs.timescale.com/latest/using-timescaledb/update-timescale/update-db-2) documentation. You will find straight-forward instructions and recommendations to ensure everything is updated and works correctly.

> # Why we’ve made these changes

Author: OK. I tried to look at other Docs examples to see what we've done previously. Maybe a long-term review at some point to bring consistency.

> It’s been two years since we released TimescaleDB 1.0 in October 2018 with the ambition to make it easy to scale PostgreSQL for time-series workloads. Since that initial release, we’ve added many new capabilities: support for compression, continuous aggregates, data lifecycle management, and numerous optimizations, as well as two new underlying PostgreSQL versions (11 and 12). At the same time, the core of TimescaleDB hasn’t changed significantly: we still have hypertables, auto-partitioning, query optimizations, and a basic user experience that many have come to appreciate because things “just work”. 

Author: Going to leave ATM as this content came from the group-reviewed Doc.

## Manually refreshing regions of continuous aggregates

TimescaleDB 2.0 removes support for `REFRESH MATERIALIZED VIEW` in favor of the new, more flexible function, [refresh_continuous_aggregate](https://docs.timescale.com/v2.0/api#refresh_continuous_aggregate), which enables a user to refresh a specific window in a continuous aggregate:

```SQL
CALL refresh_continuous_aggregate(
  continuous_aggregate => 'conditions_by_2h',
  window_start         => '2020-05-01', 
  window_end           => '2020-05-03');
```

Users can use this command explicitly; a typical use of this command is to manually refresh historical data while creating a continuous aggregate policy to aggregate new data. It is even possible to build custom continuous aggregation policies using this function within the new user-defined action framework which are then scheduled via `add_job`. 

> Note that` refresh_continuous_aggregate` only recomputes the aggregated time buckets that completely fall inside the given refresh window boundaries and are in a region that has seen changes in the underlying hypertable. Thus, if no changes have occurred in the underlying source data (that is, no data has been backfilled to the region or no updates to existing data have been made), no materialization will be performed either over that region. This behavior is similar to the continuous aggregate policy and ensures more efficient operation.

Reviewer: ```suggestion
Note that` refresh_continuous_aggregate` only recomputes the aggregated time buckets that completely fall inside the given refresh window and are in a region that has seen changes in the underlying hypertable. Thus, if no changes have occurred in the underlying source data (that is, no data has been backfilled to the region or no updates to existing data have been made), no materialization will be performed either over that region. This behavior is similar to the continuous aggregate policy and ensures more efficient operation.
```

## Using data retention and continuous aggregates together [](retention-and-caggs)

With greater power and flexibility also comes a greater responsibility to understand how features interact. Specifically, users should understand the interactions between data retention policy settings and continuous aggregate settings. 

> Before starting the upgrade to TimescaleDB 2.0, **we highly recommend checking the database log for errors related to failed retention policies that were occurring in TimescaleDB 1.x** and then either remove them or update them to be compatible with existing continuous aggregates. Any remaining retention policies that are still incompatible with the `ignore_invalidation_older_than `setting will automatically be disabled with a notice during the upgrade.

Reviewer: ```suggestion
Before starting the upgrade to TimescaleDB 2.0, **we highly recommend checking the database log for errors related to failed retention policies that were occurring in TimescaleDB 1.x** and then either removing them or updating them to be compatible with existing continuous aggregates. Any remaining retention policies that are still incompatible with the `ignore_invalidation_older_than` setting will automatically be disabled with a notice during the upgrade.
```
To maintain consistency in tense.  TIWAGOS

As an example, if a data retention policy on a hypertable is set for `drop_after => '4 weeks'`, then the policy associated with a continuous aggregate on that same hypertable should have a `start_offset` less than or equal to 4 weeks.  Similarly, any manual call to `refresh_continuous_aggregate` should likely specify a `window_start` that’s also less than the date from 4 weeks ago. 

> If not understood properly, users could overwrite existing aggregated data in continuous aggregates with empty data by recomputing data over a now-dropped time window of data, not likely the result a user was expecting. Instead, users seeking to keep only a certain time window of data in their continuous aggregate view should create a data retention policy on the continuous aggregate.

Reviewer: ```suggestion
If not understood properly, users could overwrite existing aggregated data in continuous aggregates with empty data by recomputing data over a now-dropped time window of data, not likely the result a user was expecting. Instead, users seeking to keep only a certain time window of data in their continuous aggregate view should create a data retention policy on the continuous aggregate itself.
```

**This differs significantly from how TimescaleDB 1.x handled conflicts between retention policies and continuous aggregates.**

Previously, if the continuous aggregation setting `ignore_invalidation_older_than` overlapped with data that would be dropped by a retention policy, the retention policy would silently fail.  Making the retention policy work again required users to modify settings in either the retention policy or the continuous aggregate, and even then some data wasn’t always materialized as expected. 

After upgrading to TimescaleDB 2.0, **retention policies will no longer fail due to incompatibilities with continuous aggregates** and users have to ensure that retention and continuous aggregate policies have the desired interplay. 

## Differences in the handling of backfill on continuous aggregates

> In TimescaleDB 1.x, data that was backfilled into hypertables wasn’t handled optimally; modification of any hypertable data, regardless of how old, would cause the continuous aggregate materializer job to restart from the point of the earliest backfill and then it would materialize from that point forward. Unfortunately, this could also cause the materializer to get “stuck”, since the background job only processed a limited amount of data per run, as governed by the `max_interval_per_job` setting. When this happened, one job run could potentially force the next job to start all over again. 

Reviewer: ```suggestion
In TimescaleDB 1.x, data that was backfilled into hypertables wasn’t handled optimally; modification of any hypertable data, regardless of how old, would cause the continuous aggregate materializer job to restart from the point of the earliest backfill and then materialize from that point forward. Unfortunately, this could also cause the materializer to get “stuck”, since the background job only processed a limited amount of data per run, as governed by the `max_interval_per_job` setting. When this happened, one job run could potentially force the next job to start all over again. 
```

To prevent this, the `ignore_invalidation_older_than` setting could be used to ignore backfill data older than a specified interval (e.g., ignore backfill older than 1 week), preventing the materializer from restarting if hypertable data was modified beyond this interval boundary. However, this would also stop TimescalDB 1.x from tracking changes in the underlying data beyond the `ignore_invalidation_older_than` threshold too. This meant that it was not possible to revert this setting later to a larger interval (e.g., 1 month) without potentially having a mismatch between the raw data and the aggregated data in the continuous aggregate.

In contrast, TimescaleDB 2.0 never stops tracking backfill, and to avoid materializing too much historical data, one should simply use a refresh window that does not include that region of data. The backfilled region can always be refreshed at a later time, either manually or via a policy.

> To ensure that previously ignored backfill can be refreshed after the upgrade to TimescaleDB 2.0, the upgrade process will mark the region older than the `ignore_invalidation_older_than` threshold as “requiring refresh”. This allows a manual refresh to bring a continuous aggregate up-to-date with the underlying source data. If the `ignore_invalidation_older_than` threshold was reset to a larger interval at some point, we recommend setting it back to the smaller interval prior to upgrading to ensure that all the backfill can be refreshed, if one so desires. 

Reviewer: Not sure what 'larger interval' is referring to here?

Author: See what you think of this minor change.

> Note again, however, that if backfill was previously ignored due to a retention policy on the underlying hypertable, a manual refresh of older data into a continuous aggregate could remove data when hypertable chunks have been dropped due to a data retention policy as discussed in the previous section.

Reviewer: This is slightly hard to parse?  Maybe just:
> Note again, if older data has been dropped due to a data retention policy, manual refresh of data from that time range into the continuous aggregate could remove data from the aggregate .

Or something similar

## Viewing information about continuous aggregates

In TimescaleDB 2.0, views surrounding continuous aggregates (and other policies) have been simplified and generalized.

### Changes and Additions
*   [timescaledb_information.continuous_aggregates](https://docs.timescale.com/v2.0/api#timescaledb_information-continuous_aggregate): now provides information related to the materialized view, which includes the view name and owner, the real time aggregation flag, the materialization and the view definition (the select statement defining the view)
*   [timescaledb_information.jobs](https://docs.timescale.com/v2.0/api#timescaledb_information-jobs): displays information for all policies including continuous aggregates  

> *   [Timescaledb_information.job_stats](https://docs.timescale.com/v2.0/api#timescaledb_information-jobs_stats): displays job statistics related to all jobs

Reviewer: ```suggestion
*   [timescaledb_information.job_stats](https://docs.timescale.com/v2.0/api#timescaledb_information-jobs_stats): displays job statistics related to all jobs
```

You can validate these in the proper informational views, given above.

# Other Superficial API Changes

Other minor changes were made to various APIs for greater understandability and consistency, including in the following areas.

## Data Retention

### Changes and Additions
*   [drop_chunks](https://docs.timescale.com/v2.0/api#drop_chunks): This function now requires specifying a hypertable or continuous aggregate as the first argument, and does not allow dropping chunks across all hypertables in a database.  Additionally, the arguments `cascade` and `cascade_to_materializations` were removed (and behave as if the arguments were set to `false` in earlier versions). In TimescaleDB 2.0, we instead recommend creating a separate retention policy on each continuous aggregate. 

> *   [add_retention_policy](https://docs.timescale.com/v2.0/api#add_retention_policy), [remove_retention_policy](https://docs.timescale.com/v2.0/api#remove_retention_policy):  Creating (or removing) a data retention policy now have explicit functions. Additionally, the arguments `cascade` and `cascade_to_materializations` were removed (and behave as if the arguments were set to `false` in earlier versions).

Reviewer: ```suggestion
*   [add_retention_policy](https://docs.timescale.com/v2.0/api#add_retention_policy), [remove_retention_policy](https://docs.timescale.com/v2.0/api#remove_retention_policy):  Creating (or removing) a data retention policy now has explicit functions. Additionally, the arguments `cascade` and `cascade_to_materializations` were removed (and behave as if the arguments were set to `false` in earlier versions).
```

> *   [timescaledb_information.jobs](https://docs.timescale.com/v2.0/api#jobs): General information about data retention policies are now available in the main jobs view

Reviewer: ```suggestion
*   [timescaledb_information.jobs](https://docs.timescale.com/v2.0/api#jobs): General information about data retention policies are now available in the main jobs view.
```

### Removed
*   [add_drop_chunks_policy](https://docs.timescale.com/v1.7/api#add_drop_chunks_policy): removed in favor of the explicit functions above
*   [timescaledb_information.drop_chunks_policies](https://docs.timescale.com/v1.7/api#timescaledb_information-drop_chunks_policies) view has been removed in favor of the more general jobs view.

## Compression

> ### Changes and Additions

Reviewer: Some inconsistency about whether list items have periods on the end

> *   [add_compression_policy](https://docs.timescale.com/v2.0/api#add_compression_policy), [remove_compression_policy](https://docs.timescale.com/v2.0/api#remove_compression_policy):  Creating (or removing) a compression policy now have explicit functions

Reviewer: ```suggestion
*   [add_compression_policy](https://docs.timescale.com/v2.0/api#add_compression_policy), [remove_compression_policy](https://docs.timescale.com/v2.0/api#remove_compression_policy):  Creating (or removing) a compression policy now has explicit functions
```

*   [timescaledb_information.jobs](https://docs.timescale.com/v2.0/api#timescaledb_information-jobs): General information about compression policies are now available in the main jobs view

### Removed
* [add_compress_chunk_policy](https://docs.timescale.com/v1.7/api#add_compress_chunks_policy): removed in favor of the explicit functions above
* [timescaledb_information.compressed_hypertable_stats](https://docs.timescale.com/v1.7/api#timescaledb_information-compressed_hypertable_stats): removed in favor of the new `hypertable_compression_stats(hypertable)` function linked above
* [timescaledb_information.compressed_chunk_stats](https://docs.timescale.com/v1.7/api#timescaledb_information-compressed_chunk_stats): removed in favor of the new `chunk_compression_stats(hypertable)` function linked above

## Managing policies and other jobs

TimescaleDB 2.0 introduces user-defined actions and creates a more unified jobs API. Now, jobs created by the TimescaleDB policies and for user-defined actions can be managed and viewed through a single API.

> *   `add_job:` Adds a new user-defined action to the job scheduling framework.

Reviewer: No links? Are these functions not in the API yet?
